tests/test_gradient.py

```diff
@@ -139,6 +139,143 @@ def test_gradient_equivalence(self, shape, kernel, space_order, preset, nbl, dty
         assert(np.allclose(grad_u.data, grad_v.data, rtol=tolerance, atol=tolerance))
         assert(np.allclose(grad_u.data, grad_uv.data, rtol=tolerance, atol=tolerance))
 
+    @pytest.mark.parametrize('tn', [750.])
+    @pytest.mark.parametrize('spacing', [(10, 10)])
+    @pytest.mark.parametrize("dtype, tolerance", [(np.float32, 1e-4),
+                                                  (np.float64, 1e-12)])
+    @pytest.mark.parametrize('nbl', [40])
+    @pytest.mark.parametrize('preset', ['layers-isotropic'])
+    @pytest.mark.parametrize('space_order', [4])
+    @pytest.mark.parametrize('kernel', ['OT2'])
+    @pytest.mark.parametrize('shape', [(101, 101)])
+    def test_gradient_subsampling(self, shape, kernel, space_order, preset, nbl, dtype,
```

Comment:
I said it for the previous PR but I find these tests completely unreadable with all the clutter and redefinitions. There is perfectly sable setup methods in the example for this.

That `parametrize` is also IMO way too verbose for single parameters.

Reply:
I understand, but the point is for this code to be used directly for the upcoming paper - and we would like that code to be pure Devito code (Perhaps @ggorman could help me out here), and we're reusing that code as tests for Devito. These are not meant to be unit tests - they are closer to functional tests.

Happy to take any more specific feedback that doesn't involve using the wrappers from 'examples', because I think we won't be able to use those for the paper.

Reply:
I rather have an orthogonal comment. If we really want such a big test, we _at least_ need asserts against expected norms after each op.apply, because otherwise if something goes south earlier, go figure...

Reply:
I don't like the idea of hard-coding expected norms in the test tbh but if that's the only way you see of achieving the goal you have in mind, then so be it. 

What would these norms be of? Receiver data? We could equally reinterpolate all receiver data to the same dt, and assert equality/closeness between them - this way no hard-coded norms are required.

Reply:
yeah, agree it's not ideal. Basically all we need is a mechanism to check correctness after each step (ie after each `apply`)

yes, receivers should be fine, as transitively that should (sort of...) imply correctness of the wavefield norms too